flink-ml-core/src/main/java/org/apache/flink/ml/linalg/SparseVector.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.linalg;
+
+import org.apache.flink.api.common.typeinfo.TypeInfo;
+import org.apache.flink.ml.linalg.typeinfo.SparseVectorTypeInfoFactory;
+import org.apache.flink.util.Preconditions;
+
+import java.util.Arrays;
+import java.util.Objects;
+
+/** A sparse vector of double values. */
+@TypeInfo(SparseVectorTypeInfoFactory.class)
+public class SparseVector implements Vector {
+    public final int n;
+    public final int[] indices;
+    public final double[] values;
+
+    public SparseVector(int n, int[] indices, double[] values) {
+        this.n = n;
+        this.indices = indices;
+        this.values = values;
+        checkSizeAndIndicesRange();
```

Comment:
Instead of making three passes over the vector (i.e. `checkSizeAndIndicesRange`, `isIndicesSorted` and `checkDuplicatedIndices`). Could we make just one pass to improve efficiency?

Reply:
Merging all of them might bring more time or memory cost. For example:

- `isIndicesSorted()` can return false as soon as it founds reversed indices, which means its time complexity is almost `O(1)`. Merging this process with `checkSizeAndIndicesRange()` will make this check run in `O(n)` time complexity.
- If `checkDuplicatedIndices` is executed on an unsorted array, we will need an additional set to store all history indices, which brings `O(n)` memory cost.

Except the conditions above, I agree that `checkSizeAndIndicesRange()` and `checkDuplicatedIndices` can be merged to one function. I'll change the code in this way.

Reply:
Sounds good.